history/history.c

```diff
@@ -0,0 +1,1081 @@
+    int map_size = 0, installed_count0, installed_count1;
+    struct history_delta *hd =
+        (struct history_delta *)calloc(1, sizeof(struct history_delta));
+    int *installed_ids0 = NULL, *installed_ids1 = NULL;
+
+    check_ptr(hd);
+
+    /* TODO: store size in ctx and use that */
+    rc = db_rpms_maxid(ctx->db, &map_size);
+    check_rc(rc);
+
+    installed_map = (char *)calloc(map_size, sizeof(char));
+    check_ptr(installed_map);
+
+    rc = db_play_transaction(ctx->db, trans_id0, installed_map, map_size);
+    check_rc(rc);
+
+    installed_ids0 = get_ids_from_map(installed_map, map_size, &installed_count0);
+    check_ptr(installed_ids0);
+
+    safe_free(installed_map);
+
+    installed_map = (char *)calloc(map_size, sizeof(char));
+    check_ptr(installed_map);
+
+    rc = db_play_transaction(ctx->db, trans_id1, installed_map, map_size);
+    check_rc(rc);
+
+    installed_ids1 = get_ids_from_map(installed_map, map_size, &installed_count1);
+    check_ptr(installed_ids1);
+
+    rc = diff_arrays(installed_ids1, installed_count1,
+                     installed_ids0, installed_count0,
+                     &hd->removed_ids, &hd->removed_count,
+                     &hd->added_ids, &hd->added_count);
+    check_rc(rc);
+
+error:
+    safe_free(installed_ids0);
+    safe_free(installed_ids1);
+
+    safe_free(installed_map);
+
+    if (rc) {
+        history_free_delta(hd);
+        hd = NULL;
+    }
+    return hd;
+}
+
+void history_free_nevra_map(struct history_nevra_map *hnm)
+{
+    db_free_nevra_map(hnm);
+}
+
+struct history_nevra_map *history_nevra_map(struct history_ctx *ctx)
+{
+    return db_nevra_map(ctx->db);
+}
+
+char *history_get_nevra(struct history_nevra_map *hnm, int id)
+{
+    if (id > 0 && id <= hnm->count)
+        return hnm->idmap[id - 1];
+    return NULL;
+}
+
+/* set ctx to the state at trans_id */
+int history_set_state(struct history_ctx *ctx, int trans_id)
+{
+    int rc = 0;
+    char *installed_map = NULL;
+    int map_size = 0;
+
+    rc = db_rpms_maxid(ctx->db, &map_size);
+    check_rc(rc);
+
+    installed_map = (char *)calloc(map_size, sizeof(char));
+    check_ptr(installed_map);
+
+    rc = db_play_transaction(ctx->db, trans_id, installed_map, map_size);
+    check_rc(rc);
+
+    history_set_ids_from_map(ctx, installed_map, map_size);
+
+error:
+    safe_free(installed_map);
+
+    return rc;
+}
+
+/* Update state in ctx and db by setting a baseline state.
+ * This is used to initialize the db, but can be called later to set a new baseline.
+ */
+int history_record_state(struct history_ctx *ctx)
+{
+    int rc = 0;
+    char *err_msg = NULL;
+    int trans_id = 0;
+
+    /* avoid unfinished transaction record on failure or crash */
+    rc = sqlite3_exec(ctx->db, "BEGIN TRANSACTION;", 0, 0, NULL);
+    check_db_rc(ctx->db, rc);
+
+    rc = sqlite3_exec(ctx->db,
+        "CREATE TABLE IF NOT EXISTS "
+            "transactions(Id INTEGER PRIMARY KEY AUTOINCREMENT,"
```

Comment:
all these queries should be kept in a macro

Reply:
I agree for the `CREATE` queries, and I did that in PR #327 already.